seed/seed16.go

```diff
@@ -0,0 +1,294 @@
+		}
+		return nil
+	}
+
+	batch, err := loadAssertions(assertSeedDir, checkForModel)
+	if err != nil {
+		return err
+	}
+
+	// verify we have one model assertion
+	if modelRef == nil {
+		return fmt.Errorf("seed must have a model assertion")
+	}
+
+	if err := commitTo(batch); err != nil {
+		return err
+	}
+
+	a, err := modelRef.Resolve(db.Find)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot find just added assertion %v: %v", modelRef, err)
+	}
+
+	// remember db for later use
+	s.db = db
+	s.model = a.(*asserts.Model)
+
+	return nil
+}
+
+func (s *seed16) Model() (*asserts.Model, error) {
+	if s.model == nil {
+		return nil, fmt.Errorf("internal error: model assertion unset")
+	}
+	return s.model, nil
+}
+
+func (s *seed16) addSnap(sn *Snap16, tm timings.Measurer) (*Snap, error) {
+	path := filepath.Join(s.seedDir, "snaps", sn.File)
+	seedSnap := &Snap{
+		Path: path,
+		// TODO|XXX: make sure channel is right for pinned tracks
+		Channel: sn.Channel,
+		Classic: sn.Classic,
+		DevMode: sn.DevMode,
+	}
+
+	var sideInfo snap.SideInfo
+	if sn.Unasserted {
+		sideInfo.RealName = sn.Name
+	} else {
+		var si *snap.SideInfo
+		var err error
+		timings.Run(tm, "derive-side-info", fmt.Sprintf("hash and derive side info for snap %q", sn.Name), func(nested timings.Measurer) {
+			si, err = snapasserts.DeriveSideInfo(path, s.db)
+		})
+		if asserts.IsNotFound(err) {
+			return nil, fmt.Errorf("cannot find signatures with metadata for snap %q (%q)", sn.Name, path)
+		}
+		if err != nil {
+			return nil, err
+		}
+		sideInfo = *si
+		sideInfo.Private = sn.Private
+		sideInfo.Contact = sn.Contact
+	}
+
+	seedSnap.SideInfo = &sideInfo
+
+	s.snaps = append(s.snaps, seedSnap)
+
+	return seedSnap, nil
+}
+
+func (s *seed16) LoadMeta(tm timings.Measurer) error {
+	model, err := s.Model()
+	if err != nil {
+		return err
+	}
+
+	seedYamlFile := filepath.Join(s.seedDir, "seed.yaml")
+	if !osutil.FileExists(seedYamlFile) {
+		return ErrNoMeta
+	}
+
+	seedYaml, err := ReadYaml(seedYamlFile)
+	if err != nil {
+		return err
+	}
+	yamlSnaps := seedYaml.Snaps
+
+	required := naming.NewSnapSet(model.RequiredWithEssentialSnaps())
+	seeding := make(map[string]*Snap16, len(yamlSnaps))
+	for _, sn := range yamlSnaps {
+		seeding[sn.Name] = sn
+	}
+	added := make(map[string]bool, 3)
+	classic := model.Classic()
+	_, s.usesSnapdSnap = seeding["snapd"]
+
+	baseSnap := "core"
+	classicWithSnapd := false
+	if model.Base() != "" {
+		baseSnap = model.Base()
+	}
+	if classic && s.usesSnapdSnap {
+		classicWithSnapd = true
+		// there is no system-wide base as such
+		// if there is a gadget we will install its base first though
+		baseSnap = ""
+	}
+
+	// add the essential snaps
+	addEssential := func(snapName string) (*Snap, error) {
+		// be idempotent
+		if added[snapName] {
+			return nil, nil
+		}
+		yamlSnap := seeding[snapName]
+		if yamlSnap == nil {
+			return nil, fmt.Errorf("model requires but seed is missing essential snap %q", snapName)
```

Comment:
This error message sounds a bit akward, how about "essential snap %q required by the model is missing"?